kernel: mailbox: Change a mbox testcase that condition unsuitable

KangJianX opened the pull request:
fixed: #33114 and #33120
Modify the testcase that run failed on iotdk
and nsim. This testing do not need receive thread ID when invoke
k_mbox_data_get() with NULL param. The testcase purpose is invoke
this API with NULL buffer and NULL receive_id. It will cause fatal
error if use a uninitialize receive id. 

Signed-off-by: Jian Kang <jianx.kang@intel.com>

enjiamai commented on tests/kernel/mbox/mbox_api/src/test_mbox_api.c:
```diff
@@ -433,6 +433,17 @@ void test_mbox_kdefine(void)
+ */
+void test_mbox_data_get_null(void)
+{
+	ztest_test_skip();
```
My suggestion is to write the #if-else statement into test_mbox_data_get_null() function definition, don't need to write them twice.

enjiamai commented on tests/kernel/mbox/mbox_api/src/test_mbox_api.c:
```diff
@@ -433,6 +433,17 @@ void test_mbox_kdefine(void)
+/* For iotdk and ARC/nSIM, nSIM simulates full address space of memory,
+ * so all accesses outside of CCMs are valid and access to 0x0 address
+ * doesn't generate any exception. So skip this case.
+ */
```
Please add #bug number for referencing here.

enjiamai commented:
By the way, @KangJianX , notice that we need to verify this in actually board if we have.

KangJianX commented:
@enjiamai ok, thanks

IRISZZW commented:
This testcase will also fail in ARC HS. the root cause I have explained in #33114. 

we can't `k_mbox_data_get` from a `_syncing_thread` who didn't send a msg.

`sending_thread->base.swap_data` not initialized.

I believe other platforms have similar issues with this testcase.

galak commented on tests/kernel/mbox/mbox_api/src/test_mbox_api.c:
```diff
@@ -465,6 +465,13 @@ static void thread_mbox_data_get_null(void *p1, void *p2, void *p3)
  */
 void test_mbox_data_get_null(void)
 {
+#if defined(CONFIG_CPU_ARCEM) || defined(CONFIG_CPU_CORTEX_M)
```
The comment explains why ARCEM, but why CPU_CORTEX_M here?

KangJianX replied:
For Cortex-M, it can read access to address 0x0 is generally allowed. The details in commit: f620ceacde

KangJianX commented:
@IRISZZW Yes, this testing invoke API with NULL parameter, thus the failed reason is uninitialized, but this test case purpose is NULL pointer will be caught by exception. So for some platform that not support userspace or have access read 0x0, we need skip.

IRISZZW commented:
> @IRISZZW Yes, this testing invoke API with NULL parameter, thus the failed reason is uninitialized, but this test case purpose is NULL pointer will be caught by exception. So for some platform that not support userspace or have access read 0x0, we need skip. Maybe you can find some testcases that parameter with NULL will skip some platform.

@KangJianX `k_mbox_data_get` has NULL` handle case. 
https://github.com/zephyrproject-rtos/zephyr/blob/1ce264ecdc1facb932efa53f3cbdf8fea45e9727/kernel/mailbox.c#L336-L350

KangJianX commented:
> @KangJianX I believe you have some misunderstanding.

> So we can't `k_mbox_data_get` from a `_syncing_thread` who didn't send a msg, which has unpredictable result.
> 
> We need to change this test case, not just skip it.

Yes, I know what you said. You think this testcases is not suitable, need to modify, but this testcase what you think condition not suitable is design for improve some code coverage purposely.
Whatever, I will try to change testcase and ensure coverage at same time firstly.

KangJianX committed:
tests: kernel: Change a mbox testcase that condition unsuitable

Fix issue #33114 #33120. Modify the testcase that run failed on iotdk
and nsim. This testing do not need receive thread ID when invoke
k_mbox_data_get() with NULL param. The testcase purpose is invoke
this API with NULL buffer and NULL receive_id. It will cause fatal
error if use a uninitialize receive id.

Signed-off-by: Jian Kang <jianx.kang@intel.com>

LeiW000 commented:
@IRISZZW, could you take look at the PR again? Jian already pushed another commit to address your comments.

IRISZZW commented:
> @IRISZZW, could you take look at the PR again? Jian already pushed another commit to address your comments.

this fix works, thanks.

KangJianX commented:
@nashif Could you help to review and merge this PR?
